File: backend/src/domain/models/alert.py
import enum
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.config.database import Base


class AlertType(str, enum.Enum):
    BUDGET_THRESHOLD = "budget_threshold"
    LARGE_TRANSACTION = "large_transaction"
    UNUSUAL_ACTIVITY = "unusual_activity"
    LOW_BALANCE = "low_balance"
    BILL_REMINDER = "bill_reminder"
    SAVINGS_GOAL = "savings_goal"
    PREDICTION_ALERT = "prediction_alert"


class AlertSeverity(str, enum.Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class AlertStatus(str, enum.Enum):
    UNREAD = "unread"
    READ = "read"
    DISMISSED = "dismissed"
    TRIGGERED = "triggered"


class Alert(Base):
    __tablename__ = "alerts"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    type: Mapped[AlertType] = mapped_column(Enum(AlertType), nullable=False)
    severity: Mapped[AlertSeverity] = mapped_column(
        Enum(AlertSeverity), default=AlertSeverity.INFO, nullable=False
    )
    status: Mapped[AlertStatus] = mapped_column(
        Enum(AlertStatus), default=AlertStatus.UNREAD, nullable=False
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    message: Mapped[str] = mapped_column(String(1000), nullable=False)
    reference_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    reference_id: Mapped[int | None] = mapped_column(nullable=True)
    threshold_value: Mapped[float | None] = mapped_column(nullable=True)
    current_value: Mapped[float | None] = mapped_column(nullable=True)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    user: Mapped["User"] = relationship(back_populates="alerts")

File: lib/features/analysis/domain/entities/alert.dart
enum AlertType {
  budgetThreshold,
  largeTransaction,
  unusualActivity,
  lowBalance,
  billReminder,
  savingsGoal,
  predictionAlert,
}

enum AlertSeverity { info, warning, critical }
enum AlertStatus { unread, read, dismissed, triggered }

class Alert {
  final int id;
  final int userId;
  final AlertType type;
  final AlertSeverity severity;
  final AlertStatus status;
  final String title;
  final String message;
  final String? referenceType;
  final int? referenceId;
  final double? thresholdValue;
  final double? currentValue;
  final bool isRead;
  final DateTime? readAt;
  final DateTime createdAt;

  const Alert({
    required this.id,
    required this.userId,
    required this.type,
    this.severity = AlertSeverity.info,
    this.status = AlertStatus.unread,
    required this.title,
    required this.message,
    this.referenceType,
    this.referenceId,
    this.thresholdValue,
    this.currentValue,
    this.isRead = false,
    this.readAt,
    required this.createdAt,
  });
}
